backend/analyze_tailcall_graph.ml

```diff
@@ -0,0 +1,419 @@
+      | Known_fn { name; _ } -> name
+    ;;
+
+    module Tbl = Hashtbl.Make (T)
+    module Hashset = Hashset.Make (Tbl)
+  end
+
+  module Edge = struct
+    module T = struct
+      module Label = struct
+        type t =
+          | Explicit_tail_edge
+          | Explicit_nontail_edge
+          | Inferred_tail_edge
+          | Inferred_nontail_edge
+
+        let equal t1 t2 =
+          match t1, t2 with
+          | Explicit_tail_edge, Explicit_tail_edge -> true
+          | Explicit_nontail_edge, Explicit_nontail_edge -> true
+          | Inferred_tail_edge, Inferred_tail_edge -> true
+          | Inferred_nontail_edge, Inferred_nontail_edge -> true
+          | _ -> false
+        ;;
+
+        let hash t =
+          match t with
+          | Explicit_tail_edge -> 0
+          | Explicit_nontail_edge -> 1
+          | Inferred_tail_edge -> 2
+          | Inferred_nontail_edge -> 3
+        ;;
+      end
+
+      type t =
+        { from : Vertex.t
+        ; label : Label.t
+        ; to_ : Vertex.t
+        }
+
+      let equal e1 e2 =
+        Vertex.equal e1.from e2.from
+        && Label.equal e1.label e2.label
+        && Vertex.equal e1.to_ e2.to_
+      ;;
+
+      let hash e1 = Vertex.hash e1.from + Label.hash e1.label + Vertex.hash e1.to_
+    end
+
+    include T
+    module Tbl = Hashtbl.Make (T)
+    module Hashset = Hashset.Make (Tbl)
+  end
+
+  type t =
+    { vertex_by_name : Vertex.t String.Tbl.t
+    ; adjacencies : Edge.Hashset.t Vertex.Tbl.t
+    }
+
+  let init_unknown_edges t =
+    Vertex.Tbl.replace t.adjacencies Vertex.unknown (Edge.Hashset.create 100)
+  ;;
+
+  let create () =
+    let t =
+      { vertex_by_name = String.Tbl.create 100; adjacencies = Vertex.Tbl.create 100 }
+    in
+    init_unknown_edges t;
+    t
+  ;;
+
+  let reset t =
+    String.Tbl.reset t.vertex_by_name;
+    Vertex.Tbl.reset t.adjacencies;
+    init_unknown_edges t
+  ;;
+
+  let successors t (v : Vertex.t) : Edge.Hashset.t = Vertex.Tbl.find t.adjacencies v
+
+  let find_or_add_vertex t ~(fn_name : string) : Vertex.t =
+    let vertex =
+      match String.Tbl.find_opt t.vertex_by_name fn_name with
+      | None ->
+        let id = String.Tbl.length t.vertex_by_name in
+        let vertex : Vertex.t = Known_fn { id; name = fn_name } in
+        String.Tbl.replace t.vertex_by_name fn_name vertex;
+        (* Add edge from unknown *)
+        let unknown = Vertex.unknown in
+        let edge_from_unknown : Edge.t =
+          { from = unknown; to_ = vertex; label = Explicit_tail_edge }
+        in
+        Edge.Hashset.add (successors t unknown) edge_from_unknown;
+        (* Initialize vertex's adjacency set *)
+        let edges_from_vertex = Edge.Hashset.create 10 in
+        Vertex.Tbl.replace t.adjacencies vertex edges_from_vertex;
+        vertex
+      | Some vertex -> vertex
+    in
+    vertex
+  ;;
+
+  let add_edge
+    t
+    ~(from : Vertex.t)
+    ~(to_ : Vertex.t)
+    ~(actual_position : [ `Tail | `Nontail ])
```

Comment:
Let's avoid the polymorphic variant here. I think I would prefer we either mint a new variant type or using `is_tail_position:bool`.

Reply:
Added variant type.